### PR TITLE
SALTO-3167, SALTO-3144: Verify users exist before deployment

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Verify users exist before deployment

---

We want to verify users exist in the target env before deployment of instances with user values.
Changes:
 - Change all `UserReplacer` functions to return `ElemId[]` with user values paths + make `mapping` param optional (change to user values will occur if mapping is provided)
 - Change validator 

Now that all `UserReplacer` functions are exported I'll add some specific tests for this

---
_Release Notes_: 

Zendesk:
- Users will be notified when trying to deploy an element with users that are missing from the target environment

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    return relevantInstances
+      .flatMap(instance => {
+        const missingUsers = getMissingUsers(instance, users)
+        if (_.isEmpty(missingUsers)) {
+          return []
+        }
+        return [{
+          elemID: instance.elemID,
+          severity: 'Warning',
```
Comment (tomermevorach) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:71]: Why not error? This deployment will fail no?

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          message: 'Can not change instance with user fields of users that does not exist in the target environment',
```
Comment (tomermevorach) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:72]: I'd rephrase this to:
User {user email} does not exist in the target environment

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          detailedMessage: `Can not change ${instance.elemID.name} of type ${instance.elemID.typeName}, because the instance contains the following users that doesn't exist in the target environment: ${missingUsers.join(', ')}`,
```
Comment (tomermevorach) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:73]: Rephrase to:
{type name} {instance name} include a reference to user {user email} that does not exist in the target environment.
Please manually edit the element and set an existing user email or add a user with this email to the target environment.

File: packages/zendesk-adapter/src/change_validators/missing_users.ts (hunk 2 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/src/change_validators/missing_users.ts:72]: it might be a few users

File: packages/zendesk-adapter/test/user_utils.test.ts [hunk 4]
```diff
@@ -13,81 +13,459 @@
...
-        ]
-      })
-    const users = await userUtils.getUsers(mockPaginator)
-    expect(users).toEqual(
-      [
-        { id: 1, email: 'a@a.com' },
-        { id: 2, email: 'b@b.com' },
-      ]
+
+    const sectionTranslationInstance = new InstanceElement(
```
Comment (edenhassid) [inline packages/zendesk-adapter/test/user_utils.test.ts:112]: these are the same instances that are used for the user filter test right? maybe we can share this code somehow?

**Review by edenhassid — COMMENTED** [review]
looks very good!

File: packages/zendesk-adapter/src/change_validators/missing_users.ts (hunk 1 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/src/change_validators/missing_users.ts:72]: @netama preferred it would be a warning, can u please elaborate?

File: packages/zendesk-adapter/test/user_utils.test.ts (hunk 4 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/test/user_utils.test.ts:114]: I removed it from user filter test cause actually most of it is covered in the new user_utils.test

File: packages/zendesk-adapter/src/user_utils.ts [hunk 5]
```diff
@@ -46,6 +56,157 @@ const areUsers = (values: unknown): values is User[] => {
...
+      .flatMap((condition, i) => {
+        const fieldNamesToReplace = replacerParams.fieldsToReplace.map(f => f.name)
+        const conditionValue = conditionFieldValue(condition, typeName)
+        if (!fieldNamesToReplace.includes(conditionValue)) {
+          return []
+        }
+        const valueRelativePath = replacerParams.fieldsToReplace
+          .find(f => f.name === conditionValue)?.valuePath ?? ['value']
+        const value = _.get(condition, valueRelativePath)?.toString()
+        if (value !== undefined) {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/user_utils.ts:81]: maybe it will be less confusing if we do like this. what do you think? 
```suggestion
        if (value === undefined) {
        return []
        }
```

File: packages/zendesk-adapter/src/user_utils.ts [hunk 6]
```diff
@@ -46,6 +56,157 @@ const areUsers = (values: unknown): values is User[] => {
...
+        }
+      }
+      return [valuePath]
+    })
+)
+
+const replaceRestrictionImpl = (values: Values, mapping?: Record<string, string>): string[] => {
+  const restrictionRelativePath = ['restriction', 'id']
+  const id = _.get(values, restrictionRelativePath)
+  if ((values.restriction?.type === 'User') && (id !== undefined)) {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/user_utils.ts:126]: nit: same here, we can do the negation and return right after that

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+const { createPaginator } = clientUtils
+// system options that does not contain a specific user value
+const VALID_USER_VALUES = ['current_user', 'all_agents', 'requester_id', 'assignee_id', 'requester_and_ccs']
+
+const getMissingUsers = (instance: InstanceElement, existingUsers: User[]): string[] => {
+  const userPaths = TYPE_NAME_TO_REPLACER[instance.elemID.typeName]?.(instance)
+  const missingUsers = userPaths
+    .map(path => resolvePath(instance, path))
+    .filter(userValue => !VALID_USER_VALUES.includes(userValue))
+    .filter(userValue => !(existingUsers.map(user => user.email)).includes(userValue))
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:37]: we can use a set to make the search faster. (instead of the includes)

File: packages/zendesk-adapter/test/user_utils.test.ts [hunk 8]
```diff
@@ -13,81 +13,596 @@
...
+            any: [
+              { field: 'assignee_id', operator: 'is', value: 1 },
+            ],
+          },
+          policy_metrics: [],
+        },
+      )
+      usersUtilsModule.TYPE_NAME_TO_REPLACER[
+        slaPolicyMissingValues.elemID.typeName
+      ]?.(slaPolicyMissingValues, usersMapping)
```
Comment (edenhassid) [inline packages/zendesk-adapter/test/user_utils.test.ts:590]: nit: I think you can put it all in one or two lines instead of splitting to 3

**Review by edenhassid — COMMENTED** [review]
looks great! added a few small comments that I missed before.

File: packages/zendesk-adapter/src/change_validators/missing_users.ts (hunk 7 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/src/change_validators/missing_users.ts]: good idea :)

**Review by tomermevorach — APPROVED** [review]
beautiful change validator message 😍

**Review by edenhassid — APPROVED** [review]
